### PR TITLE
[Kernel] Make kv cache write kernel work with SPMD

### PR DESCRIPTION
**Pull request by lsy323**
Add a shard_map wrapper for the kv cache write kernel

Land after https://github.com/vllm-project/tpu_commons/pull/101

Include the change in the above PR for testing

### CODE REVIEW

**Comment by lsy323** [comment]
cc @bythew3i as well somehow I can only request 1 reviewer in the PR

File: tpu_commons/kernels/ragged_kv_cache_update.py [hunk 1]
```diff
@@ -122,3 +119,46 @@ def kv_cache_update(
...
+    slices: jax.
+    Array,  # [3, slices], list of (kv_cache_start, new_kv_start, slice_len)
+    kv_cache: jax.
+    Array,  # [total_num_pages * page_size, num_combined_kv_heads, head_dim]
+    num_slices: jax.Array,  # [1]
+    *,
+    page_size: int = 32,
+    num_slices_per_block: int = 8,
+    mesh: Mesh | None = None,
+    kv_cache_pspec: P | None = None,
```
Comment (bythew3i) [inline tpu_commons/kernels/ragged_kv_cache_update.py]: Looks like only num_heads sharding is supported. Please add the comment or check to Note that!

File: tpu_commons/kernels/ragged_kv_cache_update.py [hunk 2]
```diff
@@ -122,3 +119,46 @@ def kv_cache_update(
...
+):
+    if mesh is None:
+        return _kv_cache_update(new_kv, slices, kv_cache, num_slices,
+                                page_size, num_slices_per_block)
+
+    assert kv_cache_pspec is not None, \
```
Comment (bythew3i) [inline tpu_commons/kernels/ragged_kv_cache_update.py]: Nit: user of `kv_cache_update` can pass bad `kv_cache_pspec`, prefer error throw than assert.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 1 above)
Comment (lsy323) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: Thanks @bythew3i! May I ask what would be a good way to assert only num_heads sharding is supported? I was thinking about doing something like `kv_cache_pspec == P(None, 'x', None)`, but it might fail if caller uses a different axis name, or the last `None` is omitted.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 2 above)
Comment (lsy323) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: Thanks! I just updated to raise a `ValueError` instead of assert

File: tpu_commons/models/torchax/torchax_wrapper.py [hunk 3]
```diff
@@ -193,3 +193,43 @@ def wrap_shard_map(q, kv_pages, kv_lens, page_indices, cu_q_lens,
...
+    slices: jax.
+    Array,  # [3, slices], list of (kv_cache_start, new_kv_start, slice_len)
+    kv_cache: jax.
+    Array,  # [total_num_pages * page_size, num_combined_kv_heads, head_dim]
+    num_slices: jax.Array,  # [1]
+    *,
+    page_size: int = 32,
+    num_slices_per_block: int = 8,
+) -> Array:
+    # TODO: Get rid of this wrapper and call from pallas.py directly. Need to
```
Comment (yaochengji) [inline tpu_commons/models/torchax/torchax_wrapper.py:214]: I think we can have a global context and get the mesh information from there later

Comment (lsy323) [reply tpu_commons/models/torchax/torchax_wrapper.py:214]: That's a good idea! Thanks @yaochengji, since right now the multi-chip path is broken, so let me do that properly in a separate PR.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 1 above)
Comment (bythew3i) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: I did not know the last `None` can be omitted? Are you saying `kv_cache_pspec == P(None, 'x', None)==P(None, 'x')`?

Can we check if size == 3, the kv_cache_pspec[0] and kv_cache_pspec[2] are None? If not, throw error?

Comment (lsy323) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: > I did not know the last None can be omitted?

Yes, trailing `None`s can be ommitted. Please see the following from https://docs.jax.dev/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html#namedsharding-gives-a-way-to-express-shardings-with-names

image

We can check the pspec has a length of 2 or 3, and the 2nd dim is not None.

Comment (lsy323) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: Took another thought, actually I think we should not expect the head dim is partitioned. Since it should be legal to have `num_head` dim replicated (from the kernel's perspective)

Comment (lsy323) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: @bythew3i Thanks a lot for the review!

**Comment by lsy323** [comment]
@bythew3i @yaochengji Thanks for the review!
